chore: update pull request template, add code review checklist (#2043)
codenvy/che-docs: .github/PULL_REQUEST_TEMPLATE.md

<!-- 
Please use one of the following prefixes for the title:
docs: Documentation not including procedures. Engineering review is mandatory.
procedures: Documentation including procedures. Testing procedures is mandatory. Engineering and QE review is mandatory (Engineering can review on behalf of QE). 

> chore: Choreography, release, tooling, version upgrades.

Reviewer: What does Choreography mean here?

Author: ```suggestion
chore: Routine, release, tooling, version upgrades.
```

You are right, chore is not the short form of choreography here. The conventionalcommit word for routines.

fix: Fix build, language, links, or metadata.
-->

<!-- Read our [Contribution guide](https://github.com/eclipse/che-docs/blob/master/CONTRIBUTING.adoc) before submitting a PR. -->

> ## What does this PR change?

Author: ```suggestion
## What does this pull request change?
```

> ## What does this PR do?

Reviewer: do > change

Author: ```suggestion
## What does this PR change?
```

> ## What issues does this PR fix or reference?

Author: ```suggestion
## What issues does this pull request fix or reference?
```

> ## Specify the version of the product this PR applies to

Author: ```suggestion
## Specify the version of the product this pull request applies to
```

> ## Pull Request Checklist

Reviewer: Use sentence case everywhere?

Author: ```suggestion
## Pull Request checklist
```

> The author and the reviewers validate the content of this Pull Request with this checklist, in addition to the [automated tests](code_review_checklist.adoc).

Reviewer: with this checklist > with the following checklist

Author: ```suggestion
The author and the reviewers validate the content of this Pull Request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).
```

> The author and the reviewers validate the content of this Pull Request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).

Author: ```suggestion
The author and the reviewers validate the content of this pull request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).
```

- Any procedure:
  - [ ] Successfully tested.
- Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - Propagate the URL change in:

>     - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/src/services/bootstrap/branding.constant.ts)

Reviewer 2: The link is broken

Author: Nice spot, thanks!

```suggestion
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/packages/dashboard-frontend/src/services/bootstrap/branding.constant.ts)
```

    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

> - [ ] Builds on [Eclipse Che hosted by Red Hat](https://workspaces.openshift.com).

Reviewer 2: Do we really need this?
1. Not all people do the development online
2. For external contributors it could be an additional obstacle.

Author: All the tasks don't need to be completed by the contributor. At least one of the reviewer should take care of it. Should we make it more clear?

I believe it's important to test there, as it is a good way to make sure dogfooding is never lest aside.

Reviewer: Agree that documentation testing is crucial. If this requirement helps us with getting more PRs actually tested, let's keep it in.

> - [ ] As much as possible, *`Validate language on files added or modified`* step reports no vale warnings.

Reviewer: > `* step reports no vale warnings.`

This sentence is missing something?

Author: ```suggestion
- [ ] As much as possible, the *`Validate language on files added or modified`* step reports no vale warnings.
```

> - [ ] As much as possible, the *`Validate language on files added or modified`* step reports no vale warnings.

Reviewer 2: I have a question regarding vale. Recently I had situation when vale reported error for word `passwordless`. What one should do in such a case?

Author: It means the word is not in the Vale dictionary. We have to extend the Vale Vocabulary. It is a common situation for technical terms

* Understanding Vale Vocabulary: https://docs.errata.ai/vale/vocab
* Accept file: https://github.com/eclipse/che-docs/blob/master/.vale/styles/Vocab/Che/accept.txt
* Reject file: https://github.com/eclipse/che-docs/blob/master/.vale/styles/Vocab/Che/reject.txt

Reviewer 3: can we drop `As much as possible` since it is ambiguous?
the vale check should not report any warnings

Author: ```suggestion
- [ ] the *`Validate language on files added or modified`* step reports no vale warnings.
```

> # Add here a title for your PR

Reviewer 3: why do we nee this? Every PR comes with a title naturally

Author: ```suggestion
```

# Add here a title for your PR

<!-- 
Please use one of the following prefixes for the title:
docs: Documentation not including procedures. Engineering review is mandatory.
procedures: Documentation including procedures. Testing procedures is mandatory. Engineering and QE review is mandatory (Engineering can review on behalf of QE). 
chore: Choreography, release, tooling, version upgrades.
fix: Fix build, language, links, or metadata.
-->

<!-- Read our [Contribution guide](https://github.com/eclipse/che-docs/blob/master/CONTRIBUTING.adoc) before submitting a PR. -->

## What does this PR do?

## What issues does this PR fix or reference?

## Specify the version of the product this PR applies to

## Pull Request Checklist

The author and the reviewers validate the content of this Pull Request with this checklist, in addition to the [automated tests](code_review_checklist.adoc).

- Any procedure:
  - [ ] Successfully tested.

>   - [ ] Describes a scenario already covered by QE tests, otherwise Che QE team has acknowledged an issue.

Reviewer 2: I am afraid this will lead to creation of many issue that never be resolved.
I've already been in such situation: I had to make small change in docs, but such requirement forced me to create an issue for QE team which had no value (and the issue was closed by them afterwards). Another point, even if such issues will be resolved, it will increase: 1) Time spent by QE on supporting all the tests 2) QA instances workload. In my opinion there should be some balance: only new major features should have new integration tests.

Author: I agree. This sentence is *already there*, but hardly followed.  We have to find the right formulation to have something meaningful here.

Author: ```suggestion
```

- Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - Propagate the URL change in:
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/src/services/bootstrap/branding.constant.ts)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)
- [ ] Builds on [Eclipse Che hosted by Red Hat](https://workspaces.openshift.com).

> - [ ] *`Validate language on files added or modified`* step reports no vale warnings.

Reviewer 2: To my mind this requirement is not fair. If one changes a line in a big file, this will force that person to review whole page and then, after fixing warnings / mistakes, retest the flow and create issues for QE... too much for changing a single line...

Reviewer 2: I had another experience with `vale`: it gave warnings for just using some words or phrases. But sometimes they are just necessary. I had to contact a person from docs team and usually it ended up with justification of the phrase using in that specific place...

Author: We have already fixed all vale `errors`. We have currently only 80 vale `warnings`, I believe that targeting to fix the remaining warnings is a reachable goal. We have more than 4000 `suggestions`, and new suggestions don't block PR.

Staying compliant with a style guide is hard, yes. It's like with any linter for code: it defines a very normative way of writing, and it doesn't let room for editorial disagreements about interpretation of the style guide.

Reviewer 2: > Staying compliant with a style guide is hard, yes. It's like with any linter for code: it defines a very normative way of writing, and it doesn't let room for editorial disagreements about interpretation of the style guide.

I think they are different, totally. When a linter forces developers to comply with code style it does not prevent from using language operators inside code (it will not allow arbitrary indent, but it will not block me from using, say, an array). But vale does when says that specific word or phrase is not allowed.
I don't know how to properly resolve this on technical level: to not to block and to not to allow unnecessary usage of the phrases. The best way, seems, is to have human review instead of blocking by vale.

Author: Scope of a linter

Some code linters go much deeper than indentation. Vale is very comparable to ansible-lint. It has rules to let discourage the user to use anti-patterns. From my experience, it's an incredible tool to increase the quality and consistency of Ansible playbooks, and to help new contributors learn the good practices. "As any other linter, it is opinionated." says the doc. See the rules: https://ansible-lint.readthedocs.io/en/latest/rules.html

Blocking vs. non-blocking

Current situation:

* 0 errors. Blocks the merge, since months. Time to document it.
* 80 warnings. Don't block the merge. Now that we have reached the zero errors target, the brief for reviewers is to get rid of warnings. I believe it's good to share this information in the open.
* Thousands of suggestions. Don't block merge. Need human review. Nothing changing.

As we have lived with Vale in not blocking mode for some months, I can enumerate some disadvantages:

* "check passed" didn't mean the content of the PR is fine, reviewer needed to deep dive in the logs
* more wait time for the contributor, as the review needs more work
* increasing number of errors and warnings merged
* more post-merge fixes

Now we have lived with Vale in blocking mode for some months,  can also enumerate some advantages:

Advantages of having vale block a merge for the contributor:
* reduced wait time on entry level review
* less frictions

Advantages of having vale block a merge for the reviewers:
* confidence in the "check passed" status
* less mistakes needing a post-merge fix
* less frictions
* less repetitive tasks
* more time for complex tasks: test the procedures, ensure compliance with more complex languages rules.

Rules accuracy

There is a given: we must comply to the IBM Style Guide and the Supplementary Style Guide. That's not something open to discussion.

Vale implements a subset of these rules, which are really the basics, and we can discuss the implementation in this repository:

* Severity of a rule: error should be warning or suggestion? 
* Refine a rule.

I am all for that. The rules we are using are need maturation, but they are already in a very good shape. 

But let's discuss on each rule in particular.

Author: ```suggestion
- [ ] As much as possible, *`Validate language on files added or modified`* step reports no vale warnings.
```

- [ ] *`eclipsefdn/eca`*  is successful: all contributors have signed the Eclipse Contributor Agreement  (ECA).

### Level 2

- [ ] Any procedure:
  - [ ] Successfully tested.
  - [ ] Describes a scenario already covered by QE tests, otherwise Che QE team has acknowledged an issue.
- [ ] Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - [ ] Propagate the URL change in:
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/src/services/bootstrap/branding.constant.ts)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)
- [ ] *`Build and validate PR / link checker (pull_request)`* reports no warnings.
  - [ ] *`Validate language on files added or modified`* step reports no vale warnings.
- [ ] Content is appropriate for the intended audience.
- [ ] Effective titles.
- [ ] Information is clear and concise, no ambiguity.
- [ ] No sentences with low information value.
- [ ] Sequence of information is logical.

### Level 3

> - [ ] Builds locally, and the output looks fine.

Reviewer 3: why this is not part of the automated PR check?

Author: Here we mean: locally as opposed to the CI system. This line to catch:
- build on workspaces.openshift.com
- build on a laptop using `./tools/runnerpreview.sh`

Does it need a more explicit disambiguation?

Author: ```suggestion
- [ ]  Builds on https://workspaces.openshift.com[Eclipse Che hosted by Red Hat].
- [ ]  Builds on a local environment using ./tools/runnerpreview.sh.
```

Author: Added a mandatory non automates step: Builds on https://workspaces.openshift.com[Eclipse Che hosted by Red Hat].

> - [ ] Consistent with the rest of the docs.

Reviewer 3: This part is not clear for me

Author: Language is consistent with the rest of the docs

- [ ] Downstream friendly: correct use of attributes, `ifdef` statements, examples.
- [ ] Follows modularization guidelines. See: [Modular Documentation Reference Guide](https://redhat-documentation.github.io/modular-docs/).

> - [ ] Follows the principles of minimalism. See: [The Wisdom of Crowds slides](https://docs.google.com/presentation/d/1Yeql9FrRBgKU-QlRU-nblPJ9pfZKgoKcU8SW6SQ_UqI/edit#slide=id.g1f4790d380_2_257), [The Wisdom of Crowds video](https://youtu.be/s3Em8QSXyn8), [Chunking](https://www.nngroup.com/articles/chunking/).

Reviewer 3: do we really want to have it as part of the PR check?

Author: No, moved to a dedicateddocument.

> - [ ] Language is compliant with [IBM Style Guide](https://www.oreilly.com/library/view/the-ibm-style/9780132118989/) and [Supplementary Style Guide](https://redhat-documentation.github.io/supplementary-style-guide/) rules not implemented in vale.

Reviewer 3: why this is not part of the automated PR check?

Author: "rules not implemented in vale"

Author: Because the implementation is incomplete (and will probably never be complete).

### Specify the version of the product this PR applies to.

### Pull Request Checklist for committers and reviewers

The content of this Pull Request is compliant with this checklist:

#### Level 1

> - [ ] The *eclipsefdn/eca*  chek is successful: all contributor have signed the Eclipse ECA.

Reviewer: contributors

- [ ] *Build and validate PR / link checker (pull_request)* check is successful

>     - [ ] *Build using antora* step is successful: eocumentation is building without warnings and all AsciiDoc attributes are defined.

Reviewer: eocumentation

    - [ ] *Validate links using htmltest* step is successful: internal and external links are valid.

>     - [ ] *Validate language on files added or modified* step is successful and reports no vale errors: basic language validation, no major infringements against [IBM Style Guide](https://www.oreilly.com/library/view/the-ibm-style/9780132118989/), [Red Hat Supplementary Style Guide for product documentation](https://redhat-documentation.github.io/supplementary-style-guide/), and project specific language rules.

Reviewer: no major deviations from... ?

- [ ] Documentation describes a scenario that is already covered by QE tests, otherwise an issue has been created and acknowledged by Che QE team
- [ ] Changed article references are updated where they are used (or a redirect has been configured on the docs side):
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/src/services/bootstrap/branding.constant.ts)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

#### Level 2

- [ ] *Build and validate PR / link checker (pull_request)* check reports no warnings.
    - [ ] *Validate language on files added or modified* step reports no vale warnings.
- [ ] Effective titles.
- [ ] Content is appropriate for the intended audience.
- [ ] Sequence of information is logical.
- [ ] Information is clear and concise, no ambiguity.
- [ ] No sentences with low information value.

#### Level 3 

> - [ ] Procedure successfully executed.

Reviewer: Any procedure successfully tested.
